### PR TITLE
Windows CI: Make sure that CI fails on any error

### LINKED ISSUE #39576: Windows "integration" tests report incorrect status
Seen on https://github.com/moby/moby/pull/39569#issuecomment-513266467

RS5 _failed_ due to https://github.com/moby/moby/issues/39574, but integration tests are not marked as "failed"

https://jenkins.dockerproject.org/job/Docker-PRs-WoW-RS5-Process/3055/console

```
09:13:44 Running C:\gopath\src\github.com\docker\docker\integration\container
09:13:45 INFO: Windows Base image is  microsoft/windowsservercore
09:13:45 INFO: Testing against a local daemon

...

09:13:52 === RUN   TestHealthKillContainer
09:13:59 --- FAIL: TestHealthKillContainer (7.58s)
09:13:59     health_test.go:57: assertion failed: error is not nil: Error response from daemon: Invalid signal: SIGUSR1
...

09:16:21 --- SKIP: TestWaitBlocked (0.00s)
09:16:21     wait_test.go:59: testEnv.DaemonInfo.OSType != "linux"
09:16:21 FAIL
09:16:21 Running C:\gopath\src\github.com\docker\docker\integration\image
```

Then it continues with `test-integration-cli`

```
10:22:01 OK: 524 passed, 548 skipped
10:22:01 PASS
10:22:01 ok  	github.com/docker/docker/integration-cli	3921.066s
10:22:02 INFO: Integration tests ended at 07/19/2019 10:22:02. Duration:01:05:31.5617086
```

### LINKED ISSUE #40069: CI: Windows RS5 fails on compile errors, but doesn't print the error
I ran into this a couple of times, and recalled that RS1 acts different than RS5 on these; Windows RS5 hides compilation errors, which makes it difficult to find problems in pull requests.

Here's from https://github.com/moby/moby/pull/40062:

### Failure on Windows RS5

https://ci.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-40062/runs/9/nodes/197/log/?start=0

```
?   	github.com/docker/docker/pkg/plugingetter	[no test files]
ok  	github.com/docker/docker/pkg/plugins	35.063s	coverage: 73.4% of statements

....

ERROR: make.ps1 failed:
Unit tests failed
At C:\gopath\src\github.com\docker\docker\hack\make.ps1:324 char:32
+     if ($LASTEXITCODE -ne 0) { Throw "Unit tests failed" }
```

### Same failure on Windows RS1:

https://ci.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-40062/runs/9/nodes/192/log/?start=0

```
?   	github.com/docker/docker/pkg/plugingetter	[no test files]
powershell.exe : # github.com/docker/docker/testutil/daemon
At D:\gopath\src\github.com\docker\docker@tmp\durable-9a15b210\powershellWrapper.ps1:3 char:1
+ & powershell -NoProfile -NonInteractive -ExecutionPolicy Bypass -Comm ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (# github.com/do...testutil/daemon    :String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
testutil\daemon\daemon.go:221:25: cannot use d (type *Daemon) as type string in argument to cleanupNetworkNamespace

ok  	github.com/docker/docker/pkg/plugins	35.163s	coverage: 73.4% of statements

.....

ERROR: make.ps1 failed:
Unit tests failed
At C:\gopath\src\github.com\docker\docker\hack\make.ps1:324 char:32
+     if ($LASTEXITCODE -ne 0) { Throw "Unit tests failed" }

```

**Note that _both_ report all unit tests as `ok` (passing), but Windows RS5 hides the compile error**

### PR DESCRIPTION
**Pull request by olljanat**
**- What I did**
On Windows CI tests are called using path:
Jenkins -> PowerShell -> gotestsum -> go test

gotestsum automatically return any exit code coming from go test and these changes make sure that if gotestsum exit code is not zero we exit using 1 to Jenkins too.

Also changed gotestsum to use "standard-verbose" format which is same than `-test.v` go test.

Also make sure that all outputs ends up to artifact.

**- How I did it**
Fixed PowerShell script and Jenkinsfile.

**- How to verify it**
CI passed on here and you can see that on #38469 where I also included this one broken integration test made CI failing https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-38469/19/pipeline/252 and it fails fast without need wait all tests to run.

**- A picture of a cute animal (not mandatory but encouraged)**
image

Fixes #39576
Fixes #40069

### CODE REVIEW

**Comment by olljanat** [comment]
Looks that I got this one working @thaJeztah @StefanScherer PTAL

File: Dockerfile.windows [hunk 1]
```diff
@@ -166,7 +166,7 @@ FROM microsoft/windowsservercore
 SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]
 
 ARG GO_VERSION=1.13.15
-ARG GOTESTSUM_COMMIT=v0.3.5
+ARG GOTESTSUM_COMMIT=v0.5.3
```
Comment (thaJeztah) [inline Dockerfile.windows:169]: Was this update needed to make it work?

If not, perhaps better to keep it for a separate PR, and update the linux version to the same version; https://github.com/moby/moby/blob/a70842f9c84f92741877301f33651353eddb44bb/hack/dockerfile/install/gotestsum.installer#L3

(at least a separate commit that updates both would be nice)

File: Jenkinsfile [hunk 2]
```diff
@@ -1011,17 +1011,20 @@ pipeline {
                     }
                     post {
                         always {
-                            junit testResults: 'bundles/junit-report-*.xml', allowEmptyResults: true
+                            junit testResults: 'bundles/junit-report-*.xml', allowEmptyResults: false
                             catchError(buildResult: 'SUCCESS', stageResult: 'FAILURE', message: 'Failed to create bundles.tar.gz') {
                                 powershell '''
+                                $ErrorActionPreference = 'Stop'
+                                cd $env:WORKSPACE
```
Comment (thaJeztah) [inline Jenkinsfile:1018]: interesting; is Workspace not the default directory in which Jenkins runs? (mostly curiousity from my side; not super familiar with everything Jenkins ☺️)

File: Jenkinsfile [hunk 3]
```diff
@@ -1072,17 +1075,20 @@ pipeline {
                     }
                     post {
                         always {
-                            junit testResults: 'bundles/junit-report-*.xml', allowEmptyResults: true
+                            junit testResults: 'bundles/junit-report-*.xml', allowEmptyResults: false
```
Comment (thaJeztah) [inline Jenkinsfile:1078]: IIRC, I used the "allow empty" because this step is in the "post" step, so if we failed the actual tests, we don't really care what happens next, just try to upload results if possible, and run the cleanup after

File: hack/ci/windows.ps1 [hunk 4]
```diff
@@ -789,28 +789,22 @@ Try {
...
-            }
-
-            if (Test-Path "$TEMPORIG\junit-report-unit-tests.xml") {
-                Write-Host -ForegroundColor Magenta "INFO: Unit tests results($TEMPORIG\junit-report-unit-tests.xml) exist. pwd=$pwd"
-            } else {
-                Write-Host -ForegroundColor Magenta "ERROR: Unit tests results($TEMPORIG\junit-report-unit-tests.xml) do not exist. pwd=$pwd"
-            }
-            
             # Saving where jenkins will take a look at.....
             $bundlesDir = "bundles"
```
Comment (thaJeztah) [inline hack/ci/windows.ps1:793]: Probably for a follow-up; we should get rid of some of these variables (I never was a big fan of them when they were added), and just inline the values; I think that's more readable, and doesn't require looking up what the actual path is that's set in the variable

File: hack/make.ps1 [hunk 5]
```diff
@@ -350,22 +359,18 @@ Function Run-IntegrationTests() {
         $xmlFilePath = $bundlesDir + "\junit-report-int-tests-$normDir" + ".xml"
         Set-Location $dir
         Write-Host "Running $($PWD.Path)"
+
+        $goTestArg = "--format=standard-verbose --jsonfile=$jsonFilePath --junitfile=$xmlFilePath -- $env:INTEGRATION_TESTFLAGS"
```
Comment (thaJeztah) [inline hack/make.ps1:363]: Was there a reason for moving this to an intermediate variable?

File: hack/ci/windows.ps1 (hunk 4 above)
Comment (thaJeztah) [reply hack/ci/windows.ps1:793]: Actually; if this is always "bundles" perhaps we should do it as part of this PR (if you agree with that change)

**Review by thaJeztah — COMMENTED** [review]
Thank you so much!! Not near my computer, but had an initial look at the changes; left some questions inline

**Review by StefanScherer — APPROVED** [review]
SGTM
Thanks

File: Dockerfile.windows (hunk 1 above)
Comment (olljanat) [reply Dockerfile.windows:169]: Now sure about current version. I tried earlier with pkgname-and-test-fails format and that was not available on that old version.

Comment (thaJeztah) [reply Dockerfile.windows:169]: OK, if we need the update, probably good to update both the Dockerfile.windows and Dockerfile (linux) here, so that they're on the same version 👍

File: Jenkinsfile (hunk 2 above)
Comment (olljanat) [reply Jenkinsfile:1018]: Afaiu, it is on `stages{}` but now on `post{}` phase.

File: Jenkinsfile (hunk 3 above)
Comment (olljanat) [reply Jenkinsfile:1078]: Ok. Can change back

File: hack/ci/windows.ps1 (hunk 4 above)
Comment (olljanat) [reply hack/ci/windows.ps1:793]: Sure. I can hardcode those.

File: hack/make.ps1 (hunk 5 above)
Comment (olljanat) [reply hack/make.ps1:363]: Just syncing style between `Run-UnitTests()` and `Run-IntegrationTests()` functions.

File: Jenkinsfile (hunk 2 above)
Comment (thaJeztah) [reply Jenkinsfile:1018]: Ah, that would explain. Sounds like I need to do some reading up on that, thanks!

(was wondering if we could simplify further)

File: hack/make.ps1 (hunk 5 above)
Comment (thaJeztah) [reply hack/make.ps1:363]: Gotcha; yes, there were some of these variables in those as well. Love to get rid of more of those; especially if they're only used once. I had some discussion about that in https://github.com/moby/moby/pull/39998, but in the end didn't want to block / delay merge of it

Comment (olljanat) [reply hack/make.ps1:363]: Well if we want start cleanup Windows CI then IMO we should first remove those LCOW and integration inside of container references as those scenarios are not actually used here. I quckly checked it on https://github.com/olljanat/moby/commit/630cbc6b349f569281e3a71b9c3b1f8c2a852fe3 and looks that would let us remove a lot of old logic. Anyway it better do thst on later PR.

**Comment by olljanat** [comment]
Hmm. After update RS5 CI failed to error which have nothing to do with this change (except that now we do not ignore those anymore):
```
[2020-09-17T00:26:47.921Z] === Failed
[2020-09-17T00:26:47.921Z] === FAIL: github.com/docker/docker/integration-cli TestDockerSuite/TestBuildNotVerboseFailureWithNonExistImage (5.38s)
[2020-09-17T00:26:47.921Z]     --- FAIL: TestDockerSuite/TestBuildNotVerboseFailureWithNonExistImage (5.38s)
[2020-09-17T00:26:47.921Z]         docker_cli_build_test.go:3489: Test[quiet_build_not_exists_image] expected that quiet stderr and verbose stdout are equal; quiet [Sending build context to Docker daemon  2.048kB
[2020-09-17T00:26:47.921Z] 
[2020-09-17T00:26:47.921Z]             Step 1/1 : FROM busybox11
[2020-09-17T00:26:47.921Z]             pull access denied for busybox11, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
[2020-09-17T00:26:47.921Z]             ], verbose [Sending build context to Docker daemon  2.048kB
[2020-09-17T00:26:47.921Z] 
[2020-09-17T00:26:47.921Z]             Step 1/1 : FROM busybox11
[2020-09-17T00:26:47.921Z]             Get https://registry-1.docker.io/v2/: EOF
[2020-09-17T00:26:47.921Z]             ]
[2020-09-17T00:26:47.921Z] 
[2020-09-17T00:26:47.921Z] === FAIL: github.com/docker/docker/integration-cli TestDockerSuite (2111.11s)
```
Will try what it says on next round...

File: hack/make.ps1 [hunk 6]
```diff
@@ -353,19 +362,13 @@ Function Run-IntegrationTests() {
...
         $pinfo.UseShellExecute = $false
-        $pinfo.Arguments = "--format=standard-quiet --jsonfile=$jsonFilePath --junitfile=$xmlFilePath -- $env:INTEGRATION_TESTFLAGS"
+        $pinfo.Arguments = "--format=standard-verbose --jsonfile=$jsonFilePath --junitfile=$xmlFilePath -- $env:INTEGRATION_TESTFLAGS"
         $p = New-Object System.Diagnostics.Process
         $p.StartInfo = $pinfo
         $p.Start() | Out-Null
         $p.WaitForExit()
-        $err = $p.StandardError.ReadToEnd()
-        if (($LASTEXITCODE -ne 0) -and ($err -notlike "*warning: no tests to run*")) {
-            Throw "Integration tests failed: $err"
```
Comment (thaJeztah) [inline hack/make.ps1:365]: IIUC, both stdout and stderr will now be printed directly instead of buffered because you removed `$pinfo.RedirectStandardError = $true`, so no longer needed to manually print the output correct?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
🤔 Looks like something is missing in the uploaded artifacts. Here's from another PR; https://ci-next.docker.com/public/job/moby/job/PR-41460/1/, and that contains a `windowsRS5-integration-bundles.zip`, and looks like that's missing in this PR; https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-40599/22/artifacts

File: hack/make.ps1 (hunk 6 above)
Comment (olljanat) [reply hack/make.ps1:365]: I noticed that gotestsum actually prints errors also on stdout instead of stderr so that was unnecessary redirection and because I now use `$p.ExitCode` there is no need to have that old logic anymore which I added there because of way how `go test` was printing warnings.

**Comment by olljanat** [comment]
> 🤔 Looks like something is missing in the uploaded artifacts. Here's from another PR; https://ci-next.docker.com/public/job/moby/job/PR-41460/1/, and that contains a `windowsRS5-integration-bundles.zip`, and looks like that's missing in this PR; https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-40599/22/artifacts

That is weird. Looks like that Jenkins is not running correct version of definition. End of test phase those are created to correct place:
```
[2020-09-17T07:07:10.769Z] INFO: Saving daemon under test log (d:\CI\PR-40599\22\dut.out) to bundles\CIDUT.out
[2020-09-17T07:07:10.769Z] INFO: Saving daemon under test log (d:\CI\PR-40599\22\dut.err) to bundles\CIDUT.err
```
But bundle creation is looking from them from incorrect folder:
```
[2020-09-17T07:07:49.959Z] Creating windowsRS5-integration-bundles.zip
[2020-09-17T07:07:50.422Z] powershell.exe : Compress-Archive : The path 'C:\windows\TEMP/CIDUT.out' either does not exist or is not a valid file system path.
```

and those was collected correctly earlier before I changed those `allowEmptyResults` and `allowEmptyArchive` back to true
https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-40599/20/pipeline/262

**Comment by StefanScherer** [comment]
Oh yes, I think the Jenkinsfile is protected and never run the modified version directly from a PR build.

**Comment by olljanat** [comment]
> Oh yes, I think the Jenkinsfile is protected and never run the modified version directly from a PR build.

Yep. @thaJeztah is this good now? CI looks to be 💚

**Comment by thaJeztah** [comment]
> Oh yes, I think the Jenkinsfile is protected and never run the modified version directly from a PR build.

🤦 ah, of course

**Comment by thaJeztah** [comment]
I was considering opening a draft PR to test the Jenkinsfile changes, but looking again, the changes in the Jenkinsfile are all inside `catchError(buildResult: 'SUCCESS' ...` blocks, so worst case, there's a bug, but CI would still "pass".

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
Let's merge 👍

**Comment by olljanat** [comment]
> I was considering opening a draft PR to test the Jenkinsfile changes, but looking again, the changes in the Jenkinsfile are all inside `catchError(buildResult: 'SUCCESS' ...` blocks, so worst case, there's a bug, but CI would still "pass".

Yea. Looks that it exactly what happened... Some reason it does not find bundles/CIDUT.out file https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/master/647/pipeline/231

Do you have some way to login to server and actually look where those files are?

**Comment by StefanScherer** [comment]
@olljanat It‘s a bit tricky, the Windows agents are ephemeral and get destroyed after idle time. 
I could help giving access, but we need to put a sleep somewhere to have the build job running longer. We have a 2 hour timeout in the pipeline. 
I‘m away from keyboard right now, back in 2 hours and could give a helping hand then.

**Comment by olljanat** [comment]
@StefanScherer well those Windows integration tests have been disabled, broken, etc... very long time so I guess that missing artifacts is not big thing. However as it is not merged to master you probably can check those from any pull request or master build server after those are finished.

**Comment by olljanat** [comment]
@StefanScherer Weird. Looks that https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-38469/23/artifacts/ created artifacts correctly. So either Jenkins/Jenkins nodes are someway out of sync or for some reason bundles\CIDUT.out does not get generated all the times...

**Comment by StefanScherer** [comment]
@olljanat Thanks for the two example jobs. The failing job that got the artifacts failed in line 885. If this step doesn't fail, then the script changes directory in line 891.

https://github.com/moby/moby/blob/master/hack/ci/windows.ps1#L885-L891

Maybe something like this might be the issue copying files to a relative target `bundles\CIDUT.out` and `bundles\CIDUT.err`?
I'll put some more info and investigate in a separate PR.

**Comment by StefanScherer** [comment]
@olljanat Confirmed, sometimes the script is in the wrong directory.

cidut

And the `Copy-Item  "$env:TEMP\dut.out" "bundles\CIDUT.out" -Force -ErrorAction SilentlyContinue` doesn't complain if the target folder `bundles\` doesn't exist. 😅

**Comment by StefanScherer** [comment]
Updated PR https://github.com/moby/moby/pull/41463 to go to the correct directory to collect the logs.
